sway/input/keyboard.c

```diff
@@ -534,18 +534,32 @@ static void handle_xkb_context_log(struct xkb_context *context,
 
 struct xkb_keymap *sway_keyboard_compile_keymap(struct input_config *ic,
 		char **error) {
-	struct xkb_rule_names rules = {0};
-	if (ic) {
-		input_config_fill_rule_names(ic, &rules);
-	}
-
 	struct xkb_context *context = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
 	if (!sway_assert(context, "cannot create XKB context")) {
 		return NULL;
 	}
 	xkb_context_set_user_data(context, error);
 	xkb_context_set_log_fn(context, handle_xkb_context_log);
 
+	if (ic->xkb_file) {
+		FILE *keymap_file = fopen(ic->xkb_file, "r");
+		if (!sway_assert(keymap_file, "cannot open XKB file")) {
```

Comment:
I don't think someone specifying an invalid path should cause sway to abort. This should:
- Use `sway_log_errno` to log the error
- If `err` is not `NULL`, use `malloc` and `sprintf`to pass the error back to the user
- Return `NULL`

Reply:
"If err is not NULL, use malloc and sprintf to pass the error back to the user"

I am not sure what you are referring to here?

Reply:
```c
if (err) {
	size_t len = snprintf(NULL, 0, "cannot read XKB file %s: %s",
			ic->xkb_file, strerror(errno)) + 1;
	*err = malloc(len);
	if (*err) {
		snprintf(*err, len, "cannot read XKB file %s: %s",
				ic->xkb_file, strerror(errno));
	}
}
```

This will allow for the error to be shown in swaynag or outputted by swaymsg to the user.

Reply:
Thanks for explaining this clearly. I haven't used errno's in c in a while, so excuse the possibly stupid questions. By `err` do you mean `error` from the function parameter? Should this be done in addition to sway_log_errno?

Reply:
> By err do you mean error from the function parameter?

Oops sorry, yes had the name mixed up

> Should this be done in addition to sway_log_errno?

I guess it could be an else for the `if (*error)` in case the `malloc` fails. That way it wouldn't get double logged

Reply:
Should we do something similar for "cannot close xkb file"

Reply:
Nah. Just logging that in the debug log is fine. It's not really useful from a user perspective. Failure to open could be a typo in the path or a permission issue which is much more useful to the user